glab-resolved = {base|head} support

maxice8 opened the pull request:
**Description**
<!--- Describe your changes in detail -->
This is a big one:

- Create HeadRepo in internal/glrepo, works the same as `BaseRepo` but checks `glab-resolved = head` and
  also picks the first forked repo instead of just the first repo
- vendor functions used for `glab-resolved = base` into create_mr and make them work for `glab-resolved = head`
- deprecate `--target-project` as users can just use `--repo` instead

**Related Issue**
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
Resolves #406 #393 

**How Has This Been Tested?**
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Alpine

**Types of changes**
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation
- [ ] Chore (Related to CI or Packaging to platforms)

**TODO**

- Need to re-introduce the alternate code for resolution but make it prefix with `base:` and `head:`

maxice8 committed:
feat(internal/glrepo/resolver): add HeadRepo function

maxice8 committed:
feat(command/mr/create): add helper functions for headRepo handling

resolvedHeadRepo is taken from resolvedBaseRepo used in root, it is mr
create specific as it is the only place one needs to handle more than 1
repo.

headRepoOverride is taken from baseRepoOverride used by cmdutils for the
-R option, this one instead overrides the headRepo, not the baseRepo and
thus is not required elsewhere

maxice8 committed:
feat(commands/mr/create): add HeadRepo field to CreateOpts

maxice8 committed:
feat(commands/mr/create): override HeadRepo when -H is used

maxice8 committed:
feat(commands/mr/create): use headRepo where applicable

maxice8 commented:
I think we should make a release before this

profclems commented:
Yeah, I will do the release today.

Great job! I haven't gotten time to review this PR. I will find time to go through this weekend. You can mark it as draft so we don't mistakenly merge

profclems commented:
I suggest we add this to the v2.0.0 milestone. What do you think?

maxice8 commented:
> I suggest we add this to the v2.0.0 milestone. What do you think?

Sounds good, it will certainly need a rebase and rework after #411 is merged.

maxice8 commented:
I split out some fixes for the `git.Push` stuff into #413

maxice8 committed:
feat(commands/mr/create): split headRepoRemote and baseRepoRemote

use them were appropriate, when dealing with labels we want to deal with
the baseRepo (target), when getting a tracking branch we want baseRepo,
when pushing we want headRepo

maxice8 committed:
fix(commands/mr/create): compare Commits with base repo

profclems commented:
There's one commit I'm looking for but I think you've removed it. In the commit, you removed `--target-project` and allowed it to respect  `--repo` which is literally the best thing to do.
But instead of removing the `--target-project` we can just deprecate it and keep it hidden

```go
mrCreateCmd.Flags().StringVarP(&mrCreateTargetProject, "target-project", "", "", "Add target project by id or OWNER/REPO or GROUP/NAMESPACE/REPO")
mrCreateCmd.Flags().MarkHidden("--target-project")
mrCreateCmd.Flags().MarkDeprecated("--target-project", "--target-project is deprecated. Use --repo instead")
```

maxice8 commented:
Ah I must have removed it since rebasing on top of the changes on #411

I'll re-create it, thanks for noticing.

maxice8 committed:
feat(commands/mr/create): deprecate --target-project

users can use -R|--repo instead

profclems commented on internal/glrepo/resolver.go:
```diff
@@ -141,6 +141,84 @@ func (r *ResolvedRemotes) BaseRepo(prompt bool) (Interface, error) {
+	baseName := repoNames[0]
+	if len(repoNames) > 1 {
+		err := survey.AskOne(&survey.Select{
+			Message: "Which should be the base repository (used pushing branches) for this directory?",
```
I think the message here conflicts with the prompt message of the base repository. We can use `head`
```suggestion
			Message: "Which should be the head repository (used pushing branches) for this directory?",
```

maxice8 replied:
uuugh, this is another change that was lost on the commits I dropped while rebasing.

profclems commented:
I don't see any  breaking change here 🤔. The breaking change would have resulted from removing the `--target-project` but we've fixed that. 
Correct me if I'm mistaken

profclems commented:
If there're no breaking changes, we can release this as part of v1.13.0

maxice8 commented:
> I don't see any breaking change here thinking. The breaking change would have resulted from removing the `--target-project` but we've fixed that.
> Correct me if I'm mistaken

People that have `glab-resolved` with manual resolution (a.k.a not `base` or `head`) will find that it will no longer work and they need to prefix it with `base:` or `head:`

maxice8 commented:
This change is required because we cannot distinguish the manual resolution if it is meant to refer to the `base` or `head`

maxice8 committed:
refactor(commands/mr/create): rename BaseProject to SourceProject

since TargetProject is being used to refer to what Project the Merge
Request will target it is better to use precise language in this context
and refer to the Project where the Merge Request will be created as
SourceProject

maxice8 commented:
This requires #414, I can also split more away to make it easier to work with

profclems commented:
> This requires #414, I can also split more away to make it easier to work with

Yeah that's true

profclems commented:
Since there're no breaking changes now (fixed in #414), I think we can add this to the v1.13.0 release

maxice8 commented:
Sure, I'll need to rebase after #414 is merged though.

maxice8 commented:
rebased, feel free to review and ask if anything needs to be split for easier reviewing.

profclems commented on commands/mr/create/mr_create.go:
```diff
@@ -509,3 +509,45 @@ func generateMRCompareURL(opts *CreateOpts, repo glrepo.Interface) (string, erro
+	opts.HeadRepo = func() (glrepo.Interface, error) {
+		return glrepo.FromFullName(repo)
+	}
+	newRepo, err := opts.HeadRepo()
+	if err != nil {
+		return err
+	}
+	// Initialise new http client for new repo host
+	cfg, err := f.Config()
+	if err == nil {
+		cmdutils.OverrideAPIProtocol(cfg, newRepo)
+	}
+	f.HttpClient = func() (*gitlab.Client, error) {
+		return cmdutils.LabClientFunc(newRepo.RepoHost(), cfg, false)
+	}
```
The `headRepo` is basically for pushing branches so it has nothing to do with the `HttpClient`. All API calls will be made with the `baseRepo`

```suggestion
```

maxice8 replied:
fixed, thanks.

profclems commented on commands/mr/create/mr_create.go:
```diff
@@ -329,9 +334,9 @@ func NewCmdCreate(f *cmdutils.Factory) *cobra.Command {
+				fmt.Fprintf(opts.IO.StdErr, message, utils.Cyan(opts.SourceBranch), utils.Cyan(opts.TargetBranch), baseRepo.FullName())
 
-				mr, err := api.CreateMR(labClient, repo.FullName(), mrCreateOpts)
+				mr, err := api.CreateMR(labClient, headRepo.FullName(), mrCreateOpts)
```
The merge request is created in the `baseRepo` instead

```suggestion
				mr, err := api.CreateMR(labClient, baseRepo.FullName(), mrCreateOpts)
```

maxice8 replied:
Can be either really as long as SourceProjectID and TargetProjectID point to the right places, but yes makes more sense to have it against the target repo.

maxice8 replied:
pushed with `baseRepo` instead

maxice8 replied:
Yeah it is not ok to use `baseRepo`, it fails on cross-repository branches, the merge request should be from the repository where the source branch is.

so it should be created always with `headRepo`

maxice8 replied:
Merge Request created against `Leo/aports`: https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/16206
Merge Requests created against `alpine/aports`:
       https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/16205
       https://gitlab.alpinelinux.org/alpine/aports/-/merge_requests/16204

profclems commented:
`no matching remote found` error when selecting head repo
```sh
ProfClems:  ~/.../profclems/test  |master ✓|
→ glab mr create
? Which should be the base repository (used for e.g. querying issues) for this directory? glab-cli/test
? Which should be the head repository (where branches are pushed) for this directory? profclems/test
no matching remote found


ProfClems:  ~/.../profclems/test  |master ✓|
→ git remote -v
origin  git@gitlab.com:profclems/test.git (fetch)
origin  git@gitlab.com:profclems/test.git (push)
```

It was actually set but can't use it. It prints `no matching remote found` when you run `glab mr create`
```
ProfClems:  ~/.../profclems/test  |master ✓|
→ git config -l | grep glab
remote.origin.glab-resolved=base:glab-cli/test
remote.origin.glab-resolved=head
```

Also returns this:
```sh
ProfClems:  ~/.../profclems/test  |master ✓|
→ glab mr create
expected the "[HOST/]OWNER/[NAMESPACE/]REPO" format, got "head"
```

maxice8 commented:
Unfortunately we need both remotes.

```golang
headRepoRemote, err := remotes.FindByRepo(headRepo.RepoOwner(), headRepo.RepoName())
if err != nil {
	return err
}
baseRepoRemote, err := remotes.FindByRepo(baseRepo.RepoOwner(), baseRepo.RepoName())
if err != nil {
	return err
}
```

We use `headRepoRemote` here:

```golang
if err := handlePush(opts, headRepoRemote); err != nil {
	return err
}
```

to remove it we will need to rewrite `handlePush` to push to the correct remote in another way.

We use `baseRepoRemote` here:

```golang
if opts.TargetBranch == "" {
	opts.TargetBranch, _ = git.GetDefaultBranch(baseRepoRemote.PushURL.String())
}
opts.TargetTrackingBranch = fmt.Sprintf("%s/%s", baseRepoRemote.Name, opts.TargetBranch)
```

We will need to replace them with equivalents.

and Here:

```golang
err = cmdutils.LabelsPrompt(&opts.Labels, labClient, baseRepoRemote)
if err != nil {
	return 
}
```

Same as above.

profclems commented:
We can add the remote for the fork if it is not present

```go
headRepoRemote, _ := remotes.FindByRepo(headRepo.RepoOwner(), headRepo.RepoName())
// For cases when an existing remote for the head repo will be missing:
// an existing fork was discovered by querying the API.
//
// In this case, we want to add the head repo as a new git remote so we
// can push to it.
if headRepoRemote == nil && opts.ShouldPush {
	cfg, err := opts.Config()
	if err != nil {
		return err
	}
	gitProtocol, _ := cfg.Get(headRepo.RepoHost(), "git_protocol")
	token, _ := cfg.Get(headRepo.RepoHost(), "token")
	headRepoURL := opts.SourceProject.SSHURLToRepo

	if gitProtocol != "ssh" {
		u, err := api.CurrentUser(labClient)
		if err != nil {
			return fmt.Errorf("failed to get current user info: %q", err)
		}
		remoteArgs := &glrepo.RemoteArgs{
			Protocol: gitProtocol,
			Token:    token,
			Url:      headRepo.RepoHost(),
			Username: u.Username,
		}

		headRepoURL, _ = glrepo.RemoteURL(opts.SourceProject, remoteArgs)

	}

	gitRemote, err := git.AddRemote("fork", headRepoURL)
	if err != nil {
		return fmt.Errorf("error adding remote: %w", err)
	}
	headRepoRemote = &glrepo.Remote{
		Remote: gitRemote,
		Repo: headRepo,
	}
}
```

Similar approach for when the baseRepoRemote is not present

maxice8 commented:
Sounds good to me but I'd name the remote with a `glab-` prefix as to avoid any reasonable name collisions.

profclems commented:
> Sounds good to me but I'd name the remote with a `glab-` prefix as to avoid any reasonable name collisions.

Yeah makes sense

profclems commented:
I tested and it works when baseRepoRemote (in my case) is set
```sh
ProfClems:  ~/.../profclems/test  |master ✓|
→ glab mr create
? Which should be the base repository (used for e.g. querying issues) for this directory? glab-cli/test
? Which should be the head repository (where branches are pushed) for this directory? profclems/test
? Choose a template  [Use arrows to move, type to filter]
> Open a blank merge request
```
So you can implement that

maxice8 committed:
feat(commands/mr/create): create namespaced remote if none exists

maxice8 commented:
pushed, I also created #416 which should fix a mistake done on #408

profclems reviewed (commented):
The `baseRepoRemote` is the one most likely not to have a remote. So that should be added if it does not exist as well

profclems commented:
A general function like this would do:

```go
func repoRemote(labClient *gitlab.Client, opts *CreateOpts, repo glrepo.Interface, project *gitlab.Project, remoteName string) (*glrepo.Remote, error) {
	remotes, err := opts.Remotes()
	if err != nil {
		return nil, err
	}
	repoRemote, _ := remotes.FindByRepo(repo.RepoOwner(), repo.RepoName())
	if repoRemote == nil {
		cfg, err := opts.Config()
		if err != nil {
			return nil, err
		}
		gitProtocol, _ := cfg.Get(repo.RepoHost(), "git_protocol")
		token, _ := cfg.Get(repo.RepoHost(), "token")
		repoURL := project.SSHURLToRepo

		if gitProtocol != "ssh" {
			u, err := api.CurrentUser(labClient)
			if err != nil {
				return nil, fmt.Errorf("failed to get current user info: %q", err)
			}
			remoteArgs := &glrepo.RemoteArgs{
				Protocol: gitProtocol,
				Token:    token,
				Url:      repo.RepoHost(),
				Username: u.Username,
			}

			repoURL, _ = glrepo.RemoteURL(project, remoteArgs)

		}

		gitRemote, err := git.AddRemote(remoteName, repoURL)
		if err != nil {
			return nil, fmt.Errorf("error adding remote: %w", err)
		}
		repoRemote = &glrepo.Remote{
			Remote: gitRemote,
			Repo:   repo,
		}
	}

	return repoRemote, nil
}
```

So we can use it this way:

```go
headRepoRemote, err := repoRemote(labClient, opts, headRepo, opts.SourceProject, "glab-head")
if err != nil {
	return nil
}

// assume the baseRepo is the same as headRepo			
baseRepoRemote := headRepoRemote

// if baseRepo is not the same as the headRepo set the remote
if !glrepo.IsSame(baseRepo, headRepo) {
	baseRepoRemote, err = repoRemote(labClient, opts, baseRepo, opts.TargetProject, "glab-base")
	if err != nil {
		return nil
	}
}
```
This should help because any of them could be the missing remote depending on the user's choice.
It's not tested. You can optimize it whichever you feel best.

profclems commented:
I just tested and that works.

`proclems/test` is forked from `glab-cli/test` so `glab-cli/test` becomes my baseRepo (to query issues and merge requests) and my fork, `profclems/test` is my headRepo where I have write access to push branches.
```sh
ProfClems:  ~/.../profclems/test  |master → origin ✓|
→ glab mr create
? Which should be the base repository (used for e.g. querying issues) for this directory? glab-cli/test
? Which should be the head repository (where branches are pushed) for this directory? profclems/test
? Choose a template  [Use arrows to move, type to filter]
> Open a blank merge request
```

profclems reviewed (approved):
Woohoo! 

Finally, the big change is ready

profclems committed:
Merge branch 'trunk' into glab-resolved-head

maxice8 commented:
Thanks for taking care of it, I had to go take a nap
